Make cargo aware of dwp files.

khuey opened the pull request:
When using -Csplit-debuginfo=packed on Linux, rustc will produce a dwp file. Unlike the dwo files, whose paths are embedded into the binary, there's no information in the binary to help a debugger locate a dwp file. By convention, the dwp file for `<EXE>` is given the name `<EXE>.dwp` and placed next to `<EXE>`.

When cargo hardlinks the executable file rustc put in target/debug/deps into target/debug, it also needs to hardlink the dwp file along with it. Failing to do this prevents the debugger from finding the dwp file when the binary is executed from target/debug, as there's no way for the debugger to know to look in the deps subdirectory.

The split_debuginfo option is passed down into file_types to make this possible. For cargo clean manual handling is added to match the other split_debuginfo files. bin_link_for_target also passes in None because it won't care about the dwp file.

khuey commented:
cc @davidtwco

khuey commented:
CI seems broken.

weihanglo commented:
Thank you! Looks like a duplicate of <https://github.com/rust-lang/cargo/pull/11384>? I would probably ping them to see if they want to proceed. There is also a discussion regarding artifact name conflict on Windows which is worth considering.

weihanglo commented:
r? @weihanglo

khuey commented:
Thanks for the pointer to the other PR. I hadn't seen that.

The mingw issue they ran across is actually a rustc bug: rust-lang/rust#106811.  `.dwp` should be added to the filename; it should not replace the existing extension. On Linux this isn't noticed on regular binaries but it is visible if you build a dylib. I've fixed the cargo side of that in the updated commit here.

I've also added a test along the lines you suggested in the other PR. It would be good to test the dylib behavior too but I don't know how cargo handles varying rustc versions for tests. Any test would need a version of rustc with the rustc PR above merged to pass.

I think the main difference between the PRs at this point is whether the `split_debuginfo` value is passed into `file_types` (and thus whether the DWP entry is emitted only when relevant or all the time).

khuey commented:
And the value of `should_replace_hyphens` has to match the value the binary itself had. Setting it to false unconditionally as in the other PR is wrong.

ehuss commented:
> I think the main difference between the PRs at this point is whether the `split_debuginfo` value is passed into `file_types` (and thus whether the DWP entry is emitted only when relevant or all the time).

So far cargo has generated the `FileType` list with all potential files that rustc *might* generate, and doesn't necessarily differentiate based on anything like profile settings. It is generally safe to list extra files since cargo will ignore them if rustc didn't generate them. I can't offhand think of a particular problem of differentiating, but by adding it here for just dwp that makes the code inconsistent with other platforms.

khuey commented:
Alright, done.

khuey commented:
Yay the tests finally pass.

khuey commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -462,6 +462,14 @@ impl TargetInfo {
                 })
+            } else {
+                ret.push(FileType {
+                    suffix: format!("{}.dwp", suffix),
```
.so, or .exe/.dll on mingw

khuey replied:
I've left a comment about that here, and changed this to use the implicit argument.

I did *not* change the format! I added in the implementation of cargo clean since the surrounding code doesn't use implicit arguments.

khuey commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -462,6 +462,14 @@ impl TargetInfo {
+                    prefix: prefix.clone(),
+                    flavor: FileFlavor::DebugInfo,
+                    crate_type: Some(crate_type.clone()),
+                    should_replace_hyphens: crate_type != CrateType::Bin,
```
To be honest I don't know why bins behave differently from e.g. dylibs, but what's important is that we match that behavior exactly. I've left a comment to that effect.

khuey commented:
Also even once you are satisfied with this you probably shouldn't merge it before rust-lang/rust#106811 is merged.

Gankra commented:
Oh, thanks so much for picking this up and doing it right!

khuey commented:
The rustc PRs this depends on have been merged.

weihanglo reviewed (approved):
Looks good. Thanks!

weihanglo commented:
@bors r+

bors commented:
:pushpin: Commit a393267a286bdb5fa14dc38be9c0254712ea4f74 has been approved by `weihanglo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=weihanglo a393267a286bdb5fa14dc38be9c0254712ea4f74 -->
<!-- homu: {"type":"Approved","sha":"a393267a286bdb5fa14dc38be9c0254712ea4f74","approver":"weihanglo","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit a393267a286bdb5fa14dc38be9c0254712ea4f74 with merge 99b13692dd30d19caf4bfdb81edf9e65c9af0655...
<!-- homu: {"type":"BuildStarted","head_sha":"a393267a286bdb5fa14dc38be9c0254712ea4f74","merge_sha":"99b13692dd30d19caf4bfdb81edf9e65c9af0655"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/4043589829/jobs/6953575510)
Approved by: weihanglo
Pushing 99b13692dd30d19caf4bfdb81edf9e65c9af0655 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"weihanglo","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/4043589829/jobs/6953575510"},"merge_sha":"99b13692dd30d19caf4bfdb81edf9e65c9af0655"} -->
